### DISCUSSION TITLE
GPU: Having bad input lag/stuttering? Set "Vsync Mode" to "Off"

### DISCUSSION THREAD

**Discussion by banhathome** [runelite/runelite#15337] [Ideas]
tl;dr If you're using GSync/FreeSync, set the "Vsync Mode" option in the GPU plugin to "Off".

----

Awhile back, I noticed that I was getting terrible input stuttering while playing in RuneLite. After some investigation, I realized that this was occurring because I had the "Unlock FPS" option checked and I assumed there was some glitchiness with that option so I unchecked it and went about with my life. I noticed that "Vsync Mode" defaulted to "Adaptive" and I kept it like that thinking that I needed it set to that in order to use my PC's "Adaptive Sync" i.e. GSync.

Recently, I learned that my assumption of what happened here was wrong. When "Unlock FPS" is unchecked, the GPU plugin completely ignores the "Vsync Mode" option. And it turns out that the "VSync Mode" option was what was giving me problems: Adaptive VSync interferes with GSync (and probably FreeSync as well) and it causes massive input stuttering. 

----

I've opened this discussion ticket so that we could possible have a broader discussion about Vsync. I understand why "Vsync Mode" default to "Adaptive", but considering that it may cause input stuttering/lag in some users, I think it should be defaulted to "Off". In most gamer, I think a majority of players would rather shoulder the burden of screen tearing to minimize any sort of input lag.

In addition, I think there should be better documentation on Vsync, specifically on: 

1. What does the "Adaptive" option do?
1. "Vsync Mode" value is ignored when "Unlock FPS" is disabled.

Maybe we can also include this info in the ["Troubleshooting problems with the client" wiki page](https://github.com/runelite/runelite/wiki/Troubleshooting-problems-with-the-client) since it seems like a good amount of RuneLite users experience this based on what I've seen in the Discord server.

**Comment by xxEzri** [comment]
There is a lot of confusion in the community and among developers about what exactly vsync does, what adaptive sync does and what's going on with OpenGL and the operating systems.

The Vsync options in the GPU plugin and 117HD are only used when "Unlock FPS" is checked.

Those options correspond to the following:

"Off" means that the game client will freely render frames. OS has control over frame presentation.

"On" means that the game client will render onto the back buffer, then hold it and do nothing until the monitor is starting a new refresh. OpenGL asserts control over presentation (really the OS still sits above this, but OpenGL is not giving it anything to make any decisions with..)

"Adaptive" refers to adaptive VSYNC (not adaptive sync as in the monitor tech) and means that the game will behave like On, but if a monitor refresh happens mid-render then it will temporarily disengage (behaving as Off) until the next frame render begins.

-----

In theory this vsync mode sounds like a good toggle to have, the idea of vsync holds up fine, but due to implementation problems and competition from better solutions this toggle is actually terrible and should be set to Off for probably 99% of users. The vsync toggles in OpenGL which are accessed by the GPU plugin use a very basic form of double-buffered vsync which is obsolete because it has several major problems that have since been solved and because it offers no benefit compared to those better solutions. These problems still exist in "Adaptive" mode.

What's more, the overwhelming majority of users are *already using those better solutions* and by giving OpenGL control over the frame presentation we are taking away that better solution and giving them an old, janky one that introduces unneccesary stuttering, buffer lock (CPU and GPU stalling meaning that frames are late more severely and often) and incompatibility with adaptive sync **monitors** unless further configuration is done.

This smarter frame presentation is done on the OS side and works well by default for virtually any computer that anybody can buy going back a decade plus. They use multiple optional back buffers to eliminate buffer lock and give an experience which is smoother and more reliable while at equal - or often much lower - latency. They don't ever tear. This has been a thing in some form since Windows Vista, but Win10+ has some logic implemented to take advantage of adaptive sync monitors and *improve* the presentation rather than being confused by them.

I'm not sure what's going on with Mac/ Linux or if it's different distro-to-distro, but this is a solved problem and AFAIK they're doing something similar in general.

With a Vsync option like presented in the GPU plugin, it's okay as a fallback in case you are using a niche OS which does not have this feature but i think it does way more harm than good. If you somehow don't have this kind of desktop environment then you can access the same OpenGL vsync toggle via the graphics driver, so it's not strictly neccesary and the toggle just lets people swap between A and B without restarting Runelite.

-----

Some people have expressed concern about framerates running much higher than desired without this obsolete form of double-buffered vsync putting a hard ceiling on the performance level by stalling the CPU and GPU. The issue with relying on this for performance limitation is that such stalling often happens when we *do not* want it and messes up the frame presentation.

Limiting performance is important for some users, but fundamentally vsync exists to facilitate good frame presentation and not to intentionally limit performance. Making it do so actually hurts frame presentation. A secondary system for limiting FPS (intentionally stalling the CPU only in specific, desired circumstances) can achieve the best of both worlds - performance limitation, but near-perfect frame presentation at the same time. That is already implemented in the GPU plugin.


TL;DR Put the GPU plugin's Vsync mode to "off" and if you want an FPS limit, set that. If you don't, put it to 999. Default should probably be changed. Monitor adaptive sync actually also works perfectly for me recently, but it doesn't on default configuration.

**Comment by abextm** [comment]
Most users with `Unlock FPS` enabled should have vsync set to `On` or `Adaptive`. This will give you the most consistent latency, as it locks presentation to your monitors display rate.
Disabling Vsync with unlocked fps is mostly not useful as it doesn't decrease latency any and just wastes energy.
Having Vsync off with the fps limiter set to your refresh rate makes the client have no awareness of your monitor refresh giving you effectively random latency.

Vsync + Gsync is broken because the GPU driver gets confused with GPU's refresh and the UI's refresh, which we largely cannot control. The best solution generally is to disable gsync for RuneLite.

**Reply by xxEzri** [reply]
>Most users with Unlock FPS enabled should have vsync set to On or Adaptive. This will give you the most consistent latency, as it locks presentation to your monitors display rate.
Disabling Vsync with unlocked fps is mostly not useful as it doesn't decrease latency any and just wastes energy.

Off has better frame presentation when relying on Windows to vsync in the compositor - it avoids the problem of buffer lockout (every buffer is full and thus we have nowhere to render to) and thus has fewer and less severe excursions below the frametime of a single monitor refresh, especially in situations where the frametime isn't perfectly stable.

> Having Vsync off with the fps limiter set to your refresh rate makes the client have no awareness of your monitor refresh giving you effectively random latency.

I think this is an issue inherant to desyncing the UI update rate and the render rate itself, rather than from Vsync. It exists on both configs. If you turn Vsync off, the frame will always be equal or newer than if you run the same configuration but stalled the CPU for a while before beginning to render some frames.

The problem that you're bringing up if i understand correctly is that you might start rendering right before a UI update, and thus have the old UI on your new frame - but this happens all of the time with or without frame delays, they will drift in and out of sync constantly unless they are locked together.

The best ways to limit the negative impact of that is to render at a multiple of the UI update rate (like 50, 100, 150fps rather than 60) and/or to just have a framerate so high that you'll definitely sample and display the UI update almost immediately after it's processed no matter how lucky you were.

Running that second process (frame rendering) at a higher rate reduces both the average and the maximum desync between the two. This is better explained with timeline graphics and high speed cameras TBH, but it's a common problem (known as aliasing) in all sorts of fields. Because of the desync on those rates, a 50hz UI update with a 60fps frame render can actually have *more* latency and less consistent latency than doing both of them at 50hz in lockstep - but a much higher framerate e.g. 50hz UI with 180fps almost entirely avoids that pitfall. It helps a user with a 60hz monitor, although it's a lot of cost for a gain that might not be noticed.

>Vsync + Gsync is broken because the GPU driver gets confused with GPU's refresh and the UI's refresh, which we largely cannot control. The best solution generally is to disable gsync for RuneLite.

I believe that it's due to OpenGL's old Vsync code being confused by the concept of a monitor refreshing at a non-static rate as they did not exist even conceptually when it was written. The game runs fine, better then ever in fact, so long as the OpenGL vsync is disabled. The Vsync on the Windows 10 compositor works fine. We used to have problems making Gsync even engage at all but friends and I have put hundreds of hours on it recently with an excellent experience.

**Comment by gobblesnot** [comment]
Not sure if this is relevant or not, but I found this thread due to having the same issue. What you said about disabling vSync also applies if you're using 117HD in place of the GPU plugin. For anyone that finds this who isn't so savvy, looking for a fix.

**Comment by banhathome** [comment]
I want to followup on this thread with another source of input stuttering/lag in general: the Loot Tracker plugin. Whenever I have RuneLite's Loot Tracker Side Panel open, it seems like the RuneLite client stutters for a bit every time it tries to update it after a boss KC.

**Reply by DominickCobb-rs** [reply]
Try toggling the hardware acceleration option in RuneLite (Configure) and see if it continues.
